File: python/solutions/neetcode-150/arrays-and-hashing/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start

# Solution #1: Brute Force
# Time Complexity: O(n^2)
# Space Complexity: O(1)
# class Solution:
#     def twoSum(self, nums: List[int], target: int) -> List[int]:
#         for i in range(len(nums)):
#             for j in range(i+1, len(nums)):
#                 if nums[i] + nums[j] == target:
#                     return [i, j]

# Solution #2: Use a dict of complement: index
# Time Complexity: O(n)
# Space Complexity: O(n)
class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        complement_dict = {}

        for i, value in enumerate(nums):
            complement = target - value

            if complement in complement_dict:
                return [complement_dict[complement], i]

            complement_dict[value] = i

# @lc code=end

File: go/1-two-sum.go
package main

/*
Description:
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
You may assume that each input would have exactly one solution, and you may not use the same element twice.
You can return the answer in any order.

Link: https://leetcode.com/problems/two-sum/

Solution:
Use hashmap

Time complexity: O(n)
Space complexity: O(n)
*/
func twoSum(nums []int, target int) []int {
	// Create a map to store the index of the number
	indexMap := make(map[int]int)

	// Iterate through the array
	for currIndex, currNum := range nums {
		// If the required complement is present in the map, return the index
		if requiredIdx, isPresent := indexMap[target-currNum]; isPresent {
			return []int{requiredIdx, currIndex}
		}
		// Else, store the index of the current number
		indexMap[currNum] = currIndex
	}
	// If no solution is found, return an empty array
	return []int{}
}
